src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                            _proxyHelper.ManualSettingsOnly ? _proxyHelper.ProxyBypass : Interop.WinHttp.WINHTTP_NO_PROXY_BYPASS,
+                            0);
+                        if (_sessionHandle.IsInvalid)
+                        {
+                            throw new HttpRequestException(
+                                SR.net_http_client_execution_error,
+                                WinHttpException.CreateExceptionUsingLastError());
+                        }
+
+                        _sessionHandle.DangerousAddRef(ref ignore);
+                    }
+                }
+            }
+        }
+
+        private bool GetServerCredentialsFromCache(
+            Uri uri,
+            out uint serverAuthScheme,
+            out NetworkCredential serverCredentials)
+        {
+            serverAuthScheme = 0;
+            serverCredentials = null;
+
+            NetworkCredential cred = null;
+
+            lock (_credentialCacheLock)
+            {
+                foreach (uint authScheme in s_authSchemePriorityOrder)
+                {
+                    cred = _credentialCache.GetCredential(uri, s_authSchemeStringMapping[authScheme]);
+                    if (cred != null)
+                    {
+                        serverAuthScheme = authScheme;
+                        serverCredentials = cred;
+
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials serverCredentials)
+        {
+            string authType = s_authSchemeStringMapping[authScheme];
+            Debug.Assert(!string.IsNullOrEmpty(authType));
+
+            NetworkCredential cred = serverCredentials.GetCredential(uri, authType);
+            if (cred != null)
+            {
+                lock (_credentialCacheLock)
+                {
+                    try
+                    {
+                        _credentialCache.Add(uri, authType, cred);
+                    }
+                    catch (ArgumentException)
+                    {
+                        // The credential was already added.
+                    }
+                }
+            }
+        }
+
+        private async void StartRequest(object obj)
+        {
+            RequestState state = obj as RequestState;
+            bool secureConnection = false;
+            HttpResponseMessage responseMessage = null;
+            Exception savedException = null;
+            SafeInternetHandle connectHandle = null;
+            SafeInternetHandle requestHandle = null;
+            WinHttpRequestStream requestStream = null;
+            GCHandle requestStateHandle = new GCHandle();
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.Tcs.TrySetCanceled();
```

Comment:
As a minor improvement, using the new TrySetCanceled overload, you could make this ```state.Tcs.TrySetCanceled(state.CancellationToken)```... that just helps to ensure that the token is propagated to anyone getting an OperationCanceledException later on.